### DISCUSSION TITLE
Binding `serde_json::Value` parameters json-encodes strings, leaving extra `"`

### DISCUSSION THREAD

**Discussion by rkofman** [rusqlite/rusqlite#1588] [Questions] [answered]
I'm hoping somebody here might be able to help point me in the right direction: I'm using `rusqlite` with `serde_json`. The following code is just attempting to pass-through json params directly to the execute function (yes, I understand the security implications; but it's a Tauri app and it's kind of reasonable actually 🤷):

```
pub(super) fn execute(
    connection: &Connection,
    sql: String,
    params: Option<Vec<serde_json::Value>>,
) -> Res<i32> {
    let params = params.unwrap_or_default();
    let params = params_from_iter(params);

    let mut statement = connection.prepare(&sql)?;
    let count = statement.execute(params)?;
    dbg!(statement.expanded_sql());
...
}
```

But it results in all of my string values showing up as `'"foo"'` instead of `'foo'`. Basically, every string gets surrounded with an extra `"`. I believe this is because the `to_sql` [function](https://docs.rs/rusqlite/latest/src/rusqlite/types/serde_json.rs.html#17) uses `serde_json::to_string(self)` -- which produces the string `"foo"` (json encoding of the string includes quotation marks). And when that value gets passed on, it gets saved as if the string value included the `"`s. I'm very confused on how to fix this. Is this a bug in `to_sql`? Am I using it wrong?

**Comment by gwenn** [comment]
`'foo'` is not valid JSON.
Only `'"foo"'` is valid JSON.
```sh
sqlite> SELECT json_valid('foo');
0
sqlite> SELECT json_valid('"foo"');
1
```
https://sqlite.org/json1.html#jvalid

**Reply by rkofman** [reply]
I guess I made it unclear: `json` is just the protocol on the wire. I'm trying to save `TEXT` values, `INTEGER` values, etc. The SQL I'm generating ends up being:
```sql
insert into "books" ("id", "title", "authors", "hasRead") values ('"c9d8d0cf-2890-44bf-bed2-a4474dcd7bc1"', '"Frankenstein"', '"[\"Mary Shelley\"]"', 0)
```

Where I would like (`authors` is the only field expecting json):
```sql
insert into "books" ("id", "title", "authors", "hasRead") values ('c9d8d0cf-2890-44bf-bed2-a4474dcd7bc1', 'Frankenstein', '["Mary Shelley"]', 0)
```

Are you saying that the `to_sql` of serde_json values is just for saving into `json` fields; and what I need to do is map the values into regular non-json values first if I want to save into non-json fields?

**Comment by gwenn** [comment]
`rusqlite` cannot possibly know that you don't want to persist JSON.
You may use `json_extract` (not tested):
```sql
INSERT ... VALUES (json_extract(?, '$'), ...)
```
Or check those crates:
https://github.com/rusqlite/rusqlite/issues/882#issuecomment-764822649

**Reply by rkofman** [reply, marked as answer]
Thanks for the follow-up! My current understanding is that my question is a dup of this issue: https://github.com/rusqlite/rusqlite/issues/1312.

The user intent may be ambiguous: there are two actions one could take with `serde_json::Value` as part of a sql statement:
1. Encode the data as a `json` field. In this case, the output *must* be valid json (current behavior)
2. Encode as raw data (INTEGER/REAL/TEXT). For example, when translating Json parameters into raw parameter values. (what I am trying to do, and what [alexculea](https://github.com/alexculea) seems to have been trying to do in #1312.)

My erroneous expectation was that when interacting with json fields (since they are backed by a `TEXT` store) I would be expected to manually use `serde_json::to_string()` for usecase (1); which would leave the `to_sql` implementation to handle use-case (2). This _seemed_ logical because I am thinking of the data as "encoded in JSON" and expected `to_sql` to *decode it from JSON* -- skipping the step of deserializing into an intermediate representation just to serialize back into a sql string. I did not expect the decoded value to be valid JSON. I didn't imagine that (1) could be the goal of `to_sql`.

This was a completely incorrect assumption. `rusqlite` assumes that if you are handling a `serde_json::Value`, you are treating the **JSON** itself as the value rather than as an *encoding* of a value; and what `to_sql` does is translate the _JSON_ into its string representation for DB encoding. I am guessing this is because (1) is the more common action -- (2) is only really necessary if you're building fairly general-purpose tooling and libraries. e.g. I am building a rusqlite proxy for Tauri apps.

tl;dr
An intermediate type is necessary for scenario (2) because `rusqlite` assumes (1).

**Reply by rkofman** [reply]
🤔 Since I'm not the first to make the erroneous assumption, would it be reasonable to update the documentation of `to_sql` to be more explicit about producing _valid json strings_ rather than decoding from JSON?

Happy to try to come up with a specific wording change if that's useful.

**Reply by gwenn** [reply]
> tl;dr An intermediate type is necessary for scenario (2) because `rusqlite` assumes (1).

For me, `rusqlite` doesn't assume anything but **you** assume scenario (2).
("assume" definition:  suppose to be the case, without proof)

**Reply by rkofman** [reply]
I was trying to explain how, as a user, I ended up with the wrong map for the territory. Seemed useful since at least one other user was in the same position. If that's not helpful, I apologize!

I will assume that's a declination of my suggestion for a doc change -- as my incorrect assumption is rare enough.

Thank you for your time.

**Reply by gwenn** [reply]
I was just correcting who assumed wrongly: users...
Any doc improvement is welcomed.
